azurerm/internal/services/subscription/subscription_resource.go

```diff
@@ -0,0 +1,499 @@
+	}
+
+	subscriptionName := ""
+	if subscriptionNameRaw := sub.DisplayName; subscriptionNameRaw != nil {
+		subscriptionName = *sub.DisplayName
+	}
+
+	if subscriptionName == "" || subscriptionId == "" {
+		return fmt.Errorf("one or both of Subscription Name (%q) and Subscription ID (%q) could not be determined", subscriptionName, subscriptionId)
+	}
+	// remove the alias
+	resp, err := aliasClient.Delete(ctx, id.Name)
+	if err != nil {
+		if !utils.ResponseWasNotFound(resp) {
+			return fmt.Errorf("could not delete Alias %q for Subscription %q (ID: %q): %+v", id.Name, subscriptionName, subscriptionId, err)
+		}
+	}
+
+	// Cancel the Subscription
+	if _, err := subscriptionClient.Cancel(ctx, subscriptionId); err != nil {
+		return fmt.Errorf("failed to cancel Subscription: %+v", err)
+	}
+
+	if err := waitForSubscriptionStateToSettle(ctx, meta.(*clients.Client), subscriptionId, "Cancelled", d.Timeout(schema.TimeoutDelete)); err != nil {
+		return fmt.Errorf("failed to cancel Subscription %q (Alias %q): %+v", subscriptionId, id.Name, err)
+	}
+
+	return nil
+}
+
+func waitForSubscriptionStateToSettle(ctx context.Context, clients *clients.Client, subscriptionId string, targetState string, timeout time.Duration) error {
+	stateConf := &resource.StateChangeConf{
+		Refresh: func() (result interface{}, state string, err error) {
+			status, err := clients.Subscription.Client.Get(ctx, subscriptionId)
+			return status, string(status.State), err
+		},
+		PollInterval:              20 * time.Second,
+		Timeout:                   timeout,
+		ContinuousTargetOccurence: 2,
+		Delay:                     60 * time.Second,
+	}
+	switch targetState {
+	case "Cancelled":
+		stateConf.Target = []string{
+			string(subscriptions.Disabled),
+			string(subscriptions.Warned),
+		}
+		stateConf.Pending = []string{
+			string(subscriptions.Enabled),
+		}
+
+	case "Active":
+		stateConf.Target = []string{
+			string(subscriptions.Enabled),
+		}
+		stateConf.Pending = []string{
+			string(subscriptions.Disabled),
+			string(subscriptions.Warned),
+		}
+	default:
+		return fmt.Errorf("unsupported target state %q for Subscription %q", targetState, subscriptionId)
+	}
+
+	if actual, err := stateConf.WaitForState(); err != nil {
+		sub, ok := actual.(subscriptions.Subscription)
+		if !ok {
+			return fmt.Errorf("failure in parsing response while waiting for Subscription %q to become %q: %+v", subscriptionId, targetState, err)
+		}
+		actualState := sub.State
+		return fmt.Errorf("waiting for Subscription %q to become %q, currently %q", subscriptionId, targetState, actualState)
+	}
+
+	return nil
+}
+
+func checkExistingAliases(ctx context.Context, client subscriptionAlias.AliasClient, subscriptionId string) (*string, error) {
+	aliasList, err := client.List(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not List existing Subscription Aliases")
+	}
+
+	for _, v := range *aliasList.Value {
```

Comment:
there's a panic here:

```suggestion
	if aliasList.Value == nil { return nil, err }
	for _, v := range *aliasList.Value {
```